README.md
# Monit-Agregator

Web interface for visualize the status of multiples Monit instances.

[Monit](https://mmonit.com/wiki/Monit/HowTo) is a free open source utility for managing and monitoring, processes, programs, files, directories and filesystems on a UNIX system.

You can see status of services with a webui, but only for one host.
You can see all hosts with [M/Monit](https://mmonit.com/wiki/MMonit/Setup), but it's not free and open source 😥, Monit-Agregator is a very light alternative.

## Screenshot

![screenshot](./doc/screenshot.png)

## Installation

With **cargo** :
```
cargo install monit-agregator
```

From sources :
```
git clone https://codeberg.org/thasos/monit-agregator
cd monit-agregator
cargo build --release
ls -l target/release/monit-agregator
```

## Configuration

Create a file `Settings.yaml` :

```yaml
---
ip: "0.0.0.0"
port: 3030
wait_period: 60      # refresh period for hosts status
hosts:
  - name: "host1"                                  # name displayed in webui
    url: "http://192.168.0.1:28"                   # url the binary access
    public_url: "https://monit.domain.com/host1/"  # optional
  - name: "host2"
    url: "http://192.168.0.2:28"
```

## Start

```
monit-agregator -c Settings.yaml
```

## Usage

```bash
Watchdog Monit agregator, very light M/Monit alternative

USAGE:
    monit-agregator [OPTIONS]

OPTIONS:
    -b, --bind-ip <BIND_IP>
            bind ip to listen [default: 127.0.0.1]

    -c, --config <CONFIG>
            config file path [default: Settings.yaml]

    -h, --help
            Print help information

    -l, --loglevel <LOGLEVEL>
            possible values : error warn info debug trace [default: info] [default: info]

    -p, --port <PORT>
            port to listen [default: 3030]

    -r, --refresh-period <REFRESH_PERIOD>
            sleep duration between refresh of monit status for each instance [default: 60]

    -V, --version
            Print version information
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.2] - 2023-05-09
### Changed
- Migration from Gitlab to Codeberg
- Upgrade dependencies (cargo update)

## [0.1.1] - 2022-08-28
### Added
- Disable cache html

## [0.1.0] - 2022-08-28
### Added
- Init !


src/lib.rs
use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use std::process;
use std::sync::mpsc;
use std::thread;
use warp::Filter;

pub struct BindAddress {
    pub port: u16,
    pub ip: Ipv4Addr,
}
impl BindAddress {
    pub fn default() -> BindAddress {
        BindAddress {
            port: 3030,
            ip: Ipv4Addr::new(127, 0, 0, 1),
        }
    }
    pub fn new(ip: &str, port: u16) -> BindAddress {
        let ip: Ipv4Addr = match ip.parse() {
            Ok(ip) => ip,
            Err(e) => {
                eprintln!("Error parsing IP '{}' : {}", ip, e);
                process::exit(1);
            }
        };
        BindAddress { port, ip }
    }
}

/// http server, need bind addr in conf, and hosts_urls for mapping
#[tokio::main]
pub async fn serve_http(bind_ip: Ipv4Addr, bind_port: u16, hosts_urls: Vec<(String, String)>) {
    // on vous sert quoi aujourd'hui ?
    let monit_srv = warp::path!("monit" / String).map(move |hostname: String| {
        // le move sert à hosts_urls
        println!("log: req {}", hostname);
        // return monit html in plain text
        if hostname == "favicon.ico" {
            println!("log: req favicont");
            warp::reply::html("no favico".to_string())
        } else {
            println!("log: req {}", hostname);
            let hosts_urls = hosts_urls.clone(); // faut cloner hosts_urls dans ce scope car get_monit prend l'ownership
            warp::reply::html(get_monit(hosts_urls, hostname.as_str()))
        }
    });
    // bind, serve and protect
    let bind_socket = SocketAddr::new(IpAddr::V4(bind_ip), bind_port);
    println!("log: listening on {}:{}", bind_ip, bind_port);
    warp::serve(monit_srv).run(bind_socket).await;
}

/// Spawn a thread with a blocking reqwest to the monit url
/// and return the html text response in a String
pub fn get_monit(hosts_urls: Vec<(String, String)>, hostname: &str) -> String {
    let (tx, rx) = mpsc::channel();
    let hostname = hostname.to_string();
    thread::spawn(move || {
        println!("log: searching {} in list", hostname);
        for host in hosts_urls {
            if host.0 == hostname {
                println!("log: host {} found", host.0);
                let monit_url = host.1;
                println!("log: reqwesting {}", monit_url);
                match reqwest::blocking::get(&monit_url) {
                    Ok(response) => {
                        let response = response.text();
                        //let response = reqwest::blocking::get(monit_url).unwrap().text();
                        let val = match response {
                            Ok(response) => {
                                println!("log: get monit status ok");
                                response
                            }
                            Err(_) => {
                                println!("log: get monit status ERR");
                                String::from("HTTP 500 : error from monit")
                            }
                        };
                        tx.send(val).unwrap();
                    }
                    Err(_) => {
                        let msg = format!("log: error reqwesting {}", monit_url);
                        println!("{}", msg);
                        tx.send(msg).unwrap();
                    }
                };
                // TODO si hostame pas trouvé, faut renvoyer un truc comme ça...
                //} else {
                //    println!("log: host {} not found in list", hostname);
                //    tx.send("I'm not an open proxy...".to_string()).unwrap();
            }
        }
    });
    match rx.recv() {
        Ok(body) => body,
        Err(_) => "Error reqwesting".to_string(),
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        let result = 2 + 2;
        assert_eq!(result, 4);
    }
}


src/main.rs
mod conf;
mod http_server;
mod monit_request;
mod status_generation;

#[macro_use]
extern crate log; // env_logger
use std::thread;
use tokio::sync::watch;

#[tokio::main]
async fn main() {
    // build conf
    //conf::set_loglevel();
    let conf = conf::get_conf();
    const CARGO_PKG_VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");
    info!(
        "starting up... version={}",
        CARGO_PKG_VERSION.unwrap_or("version not found")
    );

    let bind_address = conf.bind_address;
    let hosts_urls = conf.hosts_urls;
    let wait_period = conf.wait_period;

    // channel for status homepage creation
    // TODO homepage vide / build skeleton here
    let (tx, rx) = watch::channel(String::from("homepage build in progress"));

    let server_hosts_urls = hosts_urls.clone();
    thread::spawn(move || {
        http_server::serve_http(bind_address, server_hosts_urls, rx);
    });

    // https://stackoverflow.com/questions/62536566/how-can-i-create-a-tokio-runtime-inside-another-tokio-runtime-without-getting-th
    let generation_loop_status_thread = thread::spawn(move || {
        status_generation::generate_home(hosts_urls, tx, wait_period);
    })
    .join();
    if generation_loop_status_thread.is_err() {
        error!("unable to start loop generation status")
    };
}


src/http_server.rs
use crate::conf::MonitHosts;
use crate::monit_request;

use std::net::SocketAddrV4;
use tokio::sync::watch;
use warp::Filter;

#[tokio::main]
pub async fn serve_http(
    bind_address: SocketAddrV4,
    hosts_urls: Vec<MonitHosts>,
    rx: watch::Receiver<String>,
) {
    // proxy to monit instances
    let monit_proxy = warp::path!("monit" / String).map(move |subpath: String| {
        info!("client reqwested host '{}'", subpath);
        // faut cloner hosts_urls dans ce scope car get_monit prend l'ownership
        let hosts_urls = hosts_urls.clone();
        warp::reply::html(monit_request::get_monit(hosts_urls, subpath.as_str()))
    });

    // TODO use :
    //     Response::builder()
    //   .header("my-custom-header", "some-value")
    //   .body("and a custom body")
    // TODO add "refresh all" button, and print the value of refresh_period
    const CARGO_PKG_VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");
    let html_head = r#"<!DOCTYPE html>
<html lang="us">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width" />
        <title>Monit Agregator</title>
        <!-- css from w3c -->
        <link rel="stylesheet" href="css/w3.css">
        <link rel="stylesheet" href="css/w3-theme-dark-grey.css">
        <!-- disable cache -->
        <meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate">
        <meta http-equiv="Pragma" content="no-cache">
        <meta http-equiv="Expires" content="0">
    </head>
    <body class="w3-theme-dark">
            <div class="w3-panel">
            <h1>Monit-Agregator</h1>
            </div>"#;
    let html_foot = format!(
        r#"<br /><div class="w3-twothird w3-panel">version : {}<br /><a href="https://codeberg.org/thasos/monit-agregator">sources</a></div></body></html>"#,
        CARGO_PKG_VERSION.unwrap_or("version not found")
    );

    // css
    let css_w3 = include_bytes!("css/w3.css");
    let css_w3_theme = include_bytes!("css/w3-theme-dark-grey.css");
    let css = warp::path!("css" / String).map(move |subpath: String| {
        info!("client reqwested css '{}'", subpath);
        match subpath.as_str() {
            "w3.css" => warp::reply::with_header(
                String::from_utf8_lossy(css_w3),
                "content-type",
                "text/css",
            ),
            "w3-theme-dark-grey.css" => warp::reply::with_header(
                String::from_utf8_lossy(css_w3_theme),
                "content-type",
                "text/css",
            ),
            // TODO comment gérer le 404 ici ???
            _ => warp::reply::with_header(
                String::from_utf8_lossy(css_w3),
                "content-type",
                "text/css",
            ),
        }
    });

    // home
    let homepage = warp::path::end().map(move || {
        info!("homepage reqwested");
        let pretty_status = rx.borrow().to_owned();
        let html_body = format!(
            r#"
            "<div class="w3-twothird w3-panel">
                <table class="w3-table w3-centered">
                    <tr>
                        {}
                    </tr>
                </table>
            </div>"#,
            pretty_status
        );
        warp::reply::html(format!("{html_head}\n{html_body}\n{html_foot}"))
    });

    let other_paths = warp::path!(String).map(move |subpath: String| {
        // move is used by hosts_urls
        info!("client reqwested host '{}'", subpath);
        match subpath.as_str() {
            "favicon.ico" => warp::reply::html(String::from("no favicon yet\n")),
            // TODO create help page
            "help" => warp::reply::html("HEEEEEELP\n".to_owned()),
            _ => {
                // must clone hosts_urls in this scope because get_monit take ownership
                warp::reply::html(String::from("not found\n"))
            }
        }
    });

    info!("listening on {}:{}", bind_address.ip(), bind_address.port());
    let routes = monit_proxy.or(homepage).or(other_paths).or(css);
    warp::serve(routes).run(bind_address).await;
}


src/monit_request.rs
use crate::conf::MonitHosts;

use log::{error, info, warn};
use std::process;
use std::sync::mpsc::channel;
use std::thread;
use std::time::Duration;

/// Spawn a thread with a blocking reqwest to the monit url
/// and return the html text response in a String
pub fn get_monit(hosts_urls: Vec<MonitHosts>, hostname: &str) -> String {
    // simple channel for body monit response
    let (tx, rx) = channel();

    // need ownership... why ?
    let hostname = hostname.to_owned();

    // create thread for new reqwest
    thread::spawn(move || {
        // verify if host reqwested is in config
        debug!("searching '{}' in conf", &hostname);
        if hosts_urls.iter().any(|host| host.name == hostname) {
            debug!("host '{}' found in conf", &hostname);
            // TODO use macro (filter ?)
            for host in hosts_urls {
                if host.name == hostname {
                    let monit_url = host.url;
                    info!("reqwesting '{}' ({})", host.name, monit_url);

                    // let's reqwest the host !
                    let client = match reqwest::blocking::Client::builder()
                        // TODO set timeout from conf
                        .timeout(Duration::from_secs(5))
                        .build()
                    {
                        Ok(client) => client,
                        Err(_) => {
                            error!("unable to create reqwest client, this should not happen");
                            process::exit(1);
                        }
                    };
                    match client.get(&monit_url).send() {
                        Ok(response) => {
                            let response = response.text();
                            let val = match response {
                                Ok(response) => {
                                    debug!("response from monit ok, forward response body");
                                    format!("{response}\n")
                                }
                                // if monit does not return a 200 HTTP, proxy respond 500
                                Err(_) => {
                                    warn!("response from monit KO, response generic 500");
                                    format!("HTTP 500 : error from monit on host {}\n", &hostname)
                                }
                            };
                            match tx.send(val) {
                                Ok(_) => (),
                                Err(e) => {
                                    error!("thread channel error : {}", e);
                                    process::exit(1);
                                }
                            };
                        }
                        // if we cannot send http reqwest for some reason
                        Err(_) => {
                            let msg = format!("error reqwesting {monit_url}");
                            warn!("{}", msg);
                            match tx.send(msg) {
                                Ok(_) => (),
                                Err(e) => {
                                    error!("thread channel error : {}", e);
                                    process::exit(1);
                                }
                            };
                        }
                    };
                }
            }
        // host not found in config
        } else {
            tx.send(format!("no host \"{}\" found in conf", &hostname))
                .unwrap();
        }
    });

    match rx.recv() {
        Ok(body) => body,
        Err(_) => String::from("Error reqwesting\n"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    //use mockito;
    use mockito::mock;

    #[test]
    fn test_get_monit() {
        // create mock content, path, etc...
        let content_tested = "blablabla";
        let _mt = mock("GET", "/")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(content_tested)
            .create();

        // call get_monit
        let hostname = "tarsis";
        let url = &mockito::server_url();
        let hosts_urls = vec![MonitHosts {
            name: hostname.to_owned(),
            url: url.to_owned(),
            public_url: None,
        }];
        assert_eq!(
            get_monit(hosts_urls, hostname),
            format!("{content_tested}\n")
        );
    }
}


src/status_generation.rs
use crate::conf::MonitHosts;
use crate::monit_request;

use std::thread;
use std::time::Duration;
use tokio::sync::watch;

#[derive(Debug, PartialEq)]
struct MonitReport {
    nb_up: Option<i32>,
    nb_down: Option<i32>,
    nb_initialising: Option<i32>,
    nb_unmonitored: Option<i32>,
    total: Option<i32>,
    percent_up: Option<f32>,
}
impl MonitReport {
    fn get_nb_up(&self) -> String {
        match self.nb_up {
            Some(nb) => nb.to_string(),
            None => String::from("unknow"),
        }
    }
    fn get_nb_down(&self) -> String {
        match self.nb_down {
            Some(nb) => nb.to_string(),
            None => String::from("unknow"),
        }
    }
    fn get_nb_initialising(&self) -> String {
        match self.nb_initialising {
            Some(nb) => nb.to_string(),
            None => String::from("unknow"),
        }
    }
    fn get_nb_unmonitored(&self) -> String {
        match self.nb_unmonitored {
            Some(nb) => nb.to_string(),
            None => String::from("unknow"),
        }
    }
    fn get_total(&self) -> String {
        match self.total {
            Some(nb) => nb.to_string(),
            None => String::from("unknow"),
        }
    }
}

fn parse_monit_report(monit_report: &str) -> MonitReport {
    // init report
    let mut report = MonitReport {
        nb_up: None,
        nb_down: None,
        nb_initialising: None,
        nb_unmonitored: None,
        total: None,
        percent_up: None,
    };
    // feed the report
    for line in monit_report.split('\n') {
        // check if line can be parsed
        if line.contains(": ") {
            let splitted_line: Vec<&str> = line.split(':').collect(); // ["up", "            85 (100.0%)"]
            let field = splitted_line[0];
            let parsed_value = splitted_line[1].trim_start(); // "85 (100.0%)"
            let splitted_value: Vec<&str> = parsed_value.split(' ').collect(); // ["85", "(100.0%)"]
            let value = splitted_value[0];
            match field {
                "up" => {
                    report.nb_up = match value.parse() {
                        Ok(val) => Some(val),
                        Err(_) => None,
                    };
                    report.percent_up = match splitted_value[1].replace(['(', ')', '%'], "").parse()
                    {
                        Ok(val) => Some(val),
                        Err(_) => None,
                    };
                }
                "down" => {
                    report.nb_down = match value.parse() {
                        Ok(val) => Some(val),
                        Err(_) => None,
                    }
                }
                "initialising" => {
                    report.nb_initialising = match value.parse() {
                        Ok(val) => Some(val),
                        Err(_) => None,
                    }
                }
                "unmonitored" => {
                    report.nb_unmonitored = match value.parse() {
                        Ok(val) => Some(val),
                        Err(_) => None,
                    }
                }
                "total" => {
                    report.total = match value.parse() {
                        Ok(val) => Some(val),
                        Err(_) => None,
                    }
                }
                &_ => (),
            }
        }
    }
    report
}

fn display_monit_report(report: MonitReport) -> String {
    // percent_up color
    //let (color, pretty_percent_up) = match report.percent_up {
    //    Some(percent) => {
    //        let color = if percent == 100.0 { "green" } else { "red" };
    //        (color, percent.to_string())
    //    }
    //    None => ("orange", String::from("⚠️ unknow ⚠️")),
    //};
    //let highlighted_percent_up = format!("<font color={}>{}</font></h2>", color, pretty_percent_up);

    let highlighted_percent_up = match report.percent_up {
        Some(percent) => {
            if percent == 100.0 {
                format!("<font color=green>✅ {percent} ✅</font></h2>")
            } else {
                format!("<font color=red>☠️ {percent} ☠️</font></h2>")
            }
        }
        None => String::from("<font color=orange>⚠️ unknow ⚠️</font></h2>"),
    };

    // pretty print in html
    let parsed_report = format!(
        "{}
        up: {}<br />
        down: {}<br />
        initialising: {}<br />
        unmonitored: {}<br />
        total: {}",
        highlighted_percent_up,
        report.get_nb_up(),
        report.get_nb_down(),
        report.get_nb_initialising(),
        report.get_nb_unmonitored(),
        report.get_total()
    );
    parsed_report
}

#[tokio::main]
pub async fn generate_home(
    monit_hosts_list: Vec<MonitHosts>,
    tx: watch::Sender<String>,
    wait_period: Duration,
) {
    // build status from hosts urls
    let mut monit_hosts_status: Vec<MonitHosts> = Vec::new();
    for monit_host in monit_hosts_list.iter() {
        monit_hosts_status.push(MonitHosts {
            name: monit_host.name.clone(),
            url: format!("{}/_report", monit_host.url.clone()),
            public_url: monit_host.public_url.clone(),
        });
    }

    // main routine loop
    loop {
        // TODO multiple reqwest in // ?
        // generate status of monit hosts
        debug!("in loop generate_home");
        let mut pretty_status = String::new();
        for (position, host) in monit_hosts_status.clone().iter().enumerate() {
            // retrieve monit status body
            let monit_report = monit_request::get_monit(monit_hosts_status.clone(), &host.name);

            // build pretty status from all hosts
            let parsed_report = parse_monit_report(&monit_report);
            let pretty_report = display_monit_report(parsed_report);

            // replace actual status
            let _old = std::mem::replace(
                &mut monit_hosts_status[position],
                MonitHosts {
                    name: host.name.clone(),
                    url: host.url.clone(),
                    public_url: host.public_url.clone(),
                },
            );
            let url_link = match &host.public_url {
                Some(url) => url.clone(),
                None => monit_hosts_list[position].url.clone(),
            };

            // if three th, do a tr
            let rest = (position + 1) % 3;
            let is_divisible = rest == 0;
            let table_new_line = if is_divisible { "</tr>\n<tr>\n" } else { "" };

            // add a html th
            pretty_status = format!(
                "{}<th><h2><a href=\"{}\">{}</a><br />\n{}</th>\n{}",
                &pretty_status,
                url_link,
                host.name.clone(),
                pretty_report,
                table_new_line,
            );

            // incremental send
            // TODO display a message "refresh in progress" ?
            // TODO error if same host twice ⚠️⚠️⚠️⚠️⚠️⚠️
            // [2022-08-27T12:20:02Z ERROR monit_agregator::monit_request] thread channel error : sending on a closed channel
            match tx.send(pretty_status.clone()) {
                Ok(_) => debug!("homepage status sended to channel"),
                Err(_) => error!("unable to send status on channel"),
            };
        }

        thread::sleep(wait_period);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_monit_report() {
        let report = "up:            85 (100.0%)
down:           2 (0.0%)
initialising:   3 (0.0%)
unmonitored:    4 (0.0%)
total:         94 services";
        let parsed_report = MonitReport {
            nb_up: Some(85),
            nb_down: Some(2),
            nb_initialising: Some(3),
            nb_unmonitored: Some(4),
            total: Some(94),
            percent_up: Some(100.0),
        };
        assert_eq!(parse_monit_report(report), parsed_report);
    }
    #[test]
    fn test_reqwest_error() {
        let error = "error reqwesting http://6.6.6.6:666/_report";
        let parsed_report = MonitReport {
            nb_up: None,
            nb_down: None,
            nb_initialising: None,
            nb_unmonitored: None,
            total: None,
            percent_up: None,
        };
        assert_eq!(parse_monit_report(error), parsed_report);
    }
    #[test]
    fn test_parse_error() {
        // percent is not a percent here, do not panic !
        let report = "up:            85 (toto%)
down:           2 (0.0%)
initialising:   3 (0.0%)
unmonitored:    4 (0.0%)
total:         94 services";
        let parsed_report = MonitReport {
            nb_up: Some(85),
            nb_down: Some(2),
            nb_initialising: Some(3),
            nb_unmonitored: Some(4),
            total: Some(94),
            percent_up: None,
        };
        assert_eq!(parse_monit_report(report), parsed_report);
    }
}


src/conf.rs
use config::Config;
use log::LevelFilter;
use serde::{Deserialize, Serialize};
use std::env;
use std::net::{Ipv4Addr, SocketAddrV4};
use std::time::Duration;

pub struct MonitAgregatorConfig {
    pub bind_address: SocketAddrV4,
    pub hosts_urls: Vec<MonitHosts>,
    pub wait_period: Duration,
}

use clap::Parser;
#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Args {
    #[clap(
        short,
        long,
        default_value = "Settings.yaml",
        help = "config file path"
    )]
    config: String,

    #[clap(
        short,
        long,
        default_value = "info",
        help = "possible values : error warn info debug trace [default: info]"
    )]
    loglevel: LevelFilter,

    #[clap(short, long, default_value = "127.0.0.1", help = "bind ip to listen")]
    bind_ip: String,

    #[clap(short, long, default_value = "3030", help = "port to listen")]
    port: u16,

    #[clap(
        short,
        long,
        default_value = "60",
        help = "sleep duration between refresh of monit status for each instance"
    )]
    refresh_period: u64,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct MonitHosts {
    pub name: String,
    pub url: String,
    pub public_url: Option<String>,
}

pub fn get_conf() -> MonitAgregatorConfig {
    // priority is clap, then config file, then default
    let args = Args::parse();
    init_logger(args.loglevel);

    let default_ip = Ipv4Addr::new(127, 0, 0, 1);
    let default_port = 3030;
    let bind_address: SocketAddrV4;

    let default_wait_period = 60;
    let wait_period: Duration = Duration::from_secs(default_wait_period);

    let mut hosts_urls: Vec<MonitHosts> = Vec::new();

    // default config_file_path
    let config_file_path = args.config;

    // get conf from file and env
    match Config::builder()
        .add_source(config::File::with_name(&config_file_path))
        .add_source(config::Environment::with_prefix("MONAGR"))
        .build()
    {
        Ok(settings) => {
            info!("config file found : {}", config_file_path);

            // socket to liston
            let parsed_clap_ip = match args.bind_ip.parse::<Ipv4Addr>() {
                Ok(parsed_clap_ip) => parsed_clap_ip,
                Err(_) => {
                    error!("error parsing ip, set to default {:?}", default_ip);
                    default_ip
                }
            };
            let ip = if parsed_clap_ip == default_ip {
                match settings.get_string("ip") {
                    Ok(ip) => match ip.parse() {
                        Ok(parsed_ip) => parsed_ip,
                        Err(_) => {
                            error!("error parsing ip, set to default {:?}", default_ip);
                            default_ip
                        }
                    },
                    Err(_) => default_ip,
                }
            } else {
                parsed_clap_ip
            };

            let port: u16 = if args.port == default_port {
                match settings.get_int("port") {
                    Ok(port) => port as u16,
                    Err(_) => default_port,
                }
            } else {
                args.port
            };
            bind_address = SocketAddrV4::new(ip, port);

            // loop stats generation period
            let wait_period = if args.refresh_period == default_wait_period {
                match settings.get_int("wait_period") {
                    Ok(wait_period) => Duration::from_secs(wait_period as u64),
                    Err(_) => Duration::from_secs(default_wait_period),
                }
            } else {
                Duration::from_secs(args.refresh_period)
            };

            debug!("refresh monit status period : {:?}", wait_period);

            // build hosts and urls
            if let Ok(hosts_list) = settings.get_array("hosts") {
                for host in hosts_list {
                    if let Ok(host_and_urls) = host.clone().try_deserialize::<MonitHosts>() {
                        hosts_urls.push(host_and_urls);
                    };
                }
            };
        }
        Err(e) => {
            warn!("no config file found ({}), starting but do nothing", e);
            bind_address = SocketAddrV4::new(default_ip, default_port);
        }
    };
    MonitAgregatorConfig {
        bind_address,
        hosts_urls,
        wait_period,
    }
}

pub fn init_logger(loglevel: LevelFilter) {
    let mut logbuilder = env_logger::Builder::from_default_env();
    logbuilder.target(env_logger::Target::Stdout);

    // if RUST_LOG is setted, and `-l` arg is not provided or default, use RUST_LOG
    if loglevel == LevelFilter::Info {
        if env::var("RUST_LOG").is_err() {
            logbuilder.filter_level(loglevel);
        }
    } else {
        logbuilder.filter_level(loglevel);
    }

    // TODO print dependencies logs ?
    logbuilder
        .filter(Some("reqwest::"), log::LevelFilter::Off)
        .filter(Some("hyper::"), log::LevelFilter::Off)
        .filter(Some("warp::"), log::LevelFilter::Off);

    match logbuilder.try_init() {
        Ok(_) => (),
        Err(_) => eprintln!("unable to initiate pretty logging"),
    };
    debug!("loglevel : {:?}", loglevel);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_conf() {
        let conf = get_conf();
        assert_eq!(conf.bind_address.ip(), &Ipv4Addr::new(0, 0, 0, 0));
        assert_eq!(conf.bind_address.port(), 3030_u16);
    }
}
